eth/transaction_signing.go

```diff
@@ -0,0 +1,121 @@
+package eth
+
+import (
+	"encoding/hex"
+	"errors"
+
+	secp256k1 "github.com/btcsuite/btcd/btcec"
+
+	"github.com/INFURA/go-ethlibs/rlp"
+)
+
+var ErrInsufficientParams = errors.New("transaction is missing values")
+var ErrFatalCrytpo = errors.New("unable to sign Tx with private key")
+
+// Looks like there have been multiple attempts to get the Koblitz curve (secp256k1) supported in golang
+//
+// https://github.com/golang/go/pull/26873 <-- rejected
+// https://github.com/golang/go/issues/26776 <-- rejected
+// using "github.com/btcsuite/btcd/btcec"
+// other alternative is to import the C library but want to avoid that if possible
+
+func (t *Transaction) Sign(privateKey string, chainId uint64) (string, error) {
+
+	if !t.check() {
+		return "", ErrInsufficientParams
+	}
+
+	pKey, err := hex.DecodeString(privateKey)
+	if err != nil {
+		return "", err
+	}
+
+	rawTx, err := t.serialize(chainId, false)
+	if err != nil {
+		return "", err
+	}
+
+	privKey, pubKey := secp256k1.PrivKeyFromBytes(secp256k1.S256(), pKey)
+	signature, err := privKey.Sign(rawTx)
+	if err != nil {
+		return "", err
+	}
+
+	verified := signature.Verify(rawTx, pubKey)
+	if !verified {
+		panic(ErrFatalCrytpo)
+	}
+
+	// extract signature (r,s,v)
+	t.signatureValues(signature)
+	signed, err := t.serialize(chainId, true)
+	return string(signed), err
+}
+
+func (t *Transaction) serialize(chainId uint64, signature bool) ([]byte, error) {
+
+	// my brain has turned to mush, pretty sure this not the most efficient way
```

Comment:
Not sure this comment is appropriate to keep :)

Reply:
fixed! 😅